app/models/manageiq/providers/amazon/inventory/parser/cloud_manager.rb

```diff
@@ -199,12 +199,51 @@ def instances
         ),
       )
 
+      @tag_mapper = ContainerLabelTagMapping.mapper
+
+      labels = parse_labels(instance["tags"])
+      taggings = map_labels_to_tags("Vm", labels)
+      map_taggings(persister_instance, taggings)
+
       instance_hardware(persister_instance, instance, flavor)
       instance_operating_system(persister_instance, instance)
+
       vm_and_template_labels(persister_instance, instance["tags"] || [])
     end
   end
 
+  def parse_labels(entity)
+    parse_identifying_attributes(entity, 'labels')
+  end
+
+  def parse_identifying_attributes(attributes, section, source = "amazon")
+    result = []
+    return result if attributes.nil?
+      attributes.each do |tag|
+        custom_attr = {
+          :name    => tag["key"],
+          :value   => tag["value"],
+        }
+      result << custom_attr
+    end
+    result
+  end
+
+  def map_labels_to_tags(model_name, labels)
+    @tag_mapper.map_labels(model_name, labels)
+  end
+
+  # Conveniently, the tags map_labels emits are already in InventoryObject<Tag> form
+  def map_taggings(parent, tags_inventory_objects)
+    model_name = parent.inventory_collection.model_class.base_class.name
+    key = [:taggings_for, model_name]
+    collection = @inv_collections[key]
```

Comment:
there are no @inv_collections[key]

first, you need to add the inv collection as @cben  is doing it https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/162

then the collection will be here as in ```@collections```

then call it e.g. as ```persister.taggings_for_vm```

Reply:
the inventory collections should be added here https://github.com/Ladas/manageiq-providers-amazon/blob/f9123d84236a66348c4bd05d499c74f94f6d7594/app/models/manageiq/providers/amazon/inventory/persister/cloud_manager.rb#L26

Reply:
btw. ```NoMethodError: undefined method `[]' for nil:NilClass``` is because @inv_collections are nil

We have @collections here, but we should not access it directly, that is why we call it as persister.collection_name

But unless you place it on @collections[collection.name] = InventoryCollection.new..., it will not be there

Reply:
I think the taggings collections can be added in app/models/manageiq/providers/amazon/inventory_collection_default/cloud_manager.rb similar to `vm_and_template_labels` (and called from https://github.com/Ladas/manageiq-providers-amazon/blob/f9123d84236a66348c4bd05d499c74f94f6d7594/app/models/manageiq/providers/amazon/inventory/persister/cloud_manager.rb#L5)
playing with that now...

Also need to add `persistor.add_inventory_collection(@tag_mapper.tags_to_resolve_collection)` somewhere.

Reply:
@cben yes, https://github.com/Ladas/manageiq-providers-amazon/blob/f9123d84236a66348c4bd05d499c74f94f6d7594/app/models/manageiq/providers/amazon/inventory/persister/cloud_manager.rb#L5 is a good place to have shared

Then we need to add it to cloud persitor:
https://github.com/Ladas/manageiq-providers-amazon/blob/f1a67bdf880ad381231e234962284c49717f5ea7/app/models/manageiq/providers/amazon/inventory/persister/cloud_manager.rb#L3

And also to target refresh persistor:
https://github.com/Ladas/manageiq-providers-amazon/blob/f1a67bdf880ad381231e234962284c49717f5ea7/app/models/manageiq/providers/amazon/inventory/persister/target_collection.rb#L2

For targeted refresh, we will need to add targeted_arel and parent_inventory_collections. Just enhanced arel condition we already have, it will look like  https://github.com/Ladas/manageiq-providers-amazon/blob/f1a67bdf880ad381231e234962284c49717f5ea7/app/models/manageiq/providers/amazon/inventory_collection_default/cloud_manager.rb#L143

(targeted refresh specs share asserts with full refresh, so we should see it fail with missing tag)